src/game/phase/phase_cutscene.c

```diff
@@ -0,0 +1,76 @@
+#include "game/phase/phase_cutscene.h"
+
+#include "game/camera.h"
+#include "game/effects.h"
+#include "game/game.h"
+#include "game/input.h"
+#include "game/items.h"
+#include "game/lara/lara_hair.h"
+#include "game/music.h"
+#include "game/output.h"
+#include "game/shell.h"
+#include "game/sound.h"
+#include "global/types.h"
+#include "global/vars.h"
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+static const int32_t m_CinematicAnimationRate = 0x8000;
+static int32_t m_FrameCount = 0;
+
+static void Phase_Cutscene_End(void);
+static GAMEFLOW_OPTION Phase_Cutscene_Control(int32_t nframes);
+static void Game_Cutscene_Draw(void);
+
+static void Phase_Cutscene_End(void)
+{
+    Music_Stop();
+    Sound_StopAllSamples();
+}
+
+static GAMEFLOW_OPTION Phase_Cutscene_Control(int32_t nframes)
+{
+    m_FrameCount += m_CinematicAnimationRate * nframes;
+    while (m_FrameCount >= 0) {
+        if (g_CineFrame >= g_NumCineFrames - 1) {
+            g_LevelComplete = true;
+            return GF_LEVEL_COMPLETE | g_CurrentLevel;
+        }
+
+        Input_Update();
+        Shell_ProcessInput();
+        Game_ProcessInput();
+
+        if (g_InputDB.menu_confirm || g_InputDB.menu_back) {
+            g_LevelComplete = true;
+            return GF_LEVEL_COMPLETE | g_CurrentLevel;
+        }
+
+        Item_Control();
+        Effect_Control();
+
+        Lara_Hair_Control();
+
+        Camera_UpdateCutscene();
+
+        g_CineFrame++;
+        m_FrameCount -= 0x10000;
+    }
+
+    return GF_NOP;
+}
+
+static void Phase_Cutscene_Draw(void)
+{
+    Game_DrawScene(true);
+    Output_AnimateTextures(g_Camera.number_frames);
+}
+
+PHASER g_CutscenePhaser = {
+    .start = NULL,
```

Comment:
Why is `Game_Cutscene_Start` with `Game_Cutscene_InitialiseHair` not used for a phaser `.start` function?

Reply:
Thanks, good catch. Some of the cutscenes need to have their camera positions hardcoded. To achieve this, we split the cutscene loading process into two gameflow sequences:

- **Start**, which initializes everything (the level, objects and whatnot),
- **Loop**, which runs the actual cutscene. This is the part got converted to a phase.
- **End**, which stops the music. This got moved to the phase .end.

This way, we get to inject extra sequences like this:

- Start,
- **Set camera X/Y/Z**,
- **Set camera angle**,
- **Play synced audio**,
- **Mesh swap**,
- Loop,
- [End] (which is now a part of `Phase_End`)

At first I didn't move `Game_Cutscene_Start` into the phase because I wanted to keep `Phase_Set` right before the `Phase_Loop`. However, it's why the music broke. The chronology was like this - for the first cutscene everything worked fine, but for the following one:

- Run GF `start_cine` sequence – load level, objects…
- Run GF camera, audio and mesh swap sequences… + **play synced audio**
- Run GF `loop_cine` seuqnece:
    - `Phase_Set(PHASE_CUTSCENE)`
        - End the previous phase (`Phase_Cutscene_End()` = **`Music_Stop()`**)
        - Start the new phase (do nothing as there was not `Phase_Cutscene_Start()`)
    - `Game_Loop()` (play the cutscene)

 I've fixed this by moving `Game_Cutscene_Start` to `Phase_Cutscene_Start`, and calling `Phase_Set` upon `start_cine`, while keeping `Game_Loop` in the `loop_cine`. This way the above flow gets changed to:

- Run GF `start_cine` sequence:
     - `Phase_Set(PHASE_CUTSCENE, level_num)`
          - `Phase_Cutscene_End()` = `Music_Stop()`
          - `Phase_Cutscene_Start()` = load level, objects…
- Run GF camera, audio and mesh swap sequences… + **play synced audio**
- Run GF `loop_cine` sequence:
    - `Game_Loop()` (play the cutscene)